Re-do commit 583ef70 on current master

dgrcode opened the pull request:
The way to use `<FormattedMessage />` with rich text:

```jsx
<FormattedMessage
  id="message.id"
  defaultMessage="Hello <big>{name}</big>"
  values={{
    name: 'Foo',
    big: chunks => <h1>{chunks}</h1>
  }}
```
Note there are two things to interpolate: the things in `{}`, and the things in `<>`.

Something I found out is that a missing value will show the default message. For example:
```jsx
<FormattedMessage
  id="message.id"
  defaultMessage="Hello {name}"
  values={{
    name: 'Foo',
  }}
```
and en.json:
```
{
  "message.id": "Hello <h1>{name}</h1>"
}
```
will show the default message because it doesn't know how to resolve the `h1` reference.

dgrcode commented:
I'd go through each component and replace plain texts with the proper `<FormatMessage />` components.

A bit more tricky can be to find strings outside of the returned jsx. I'd look for `aria-label` in the whole repo. Calls to `toast` or `useToast` can also be tricky. In that case, both `title` and `description` props accept react components ([source](https://github.com/chakra-ui/chakra-ui/blob/c14615106066349955b0eaa7b78d4f3a59a08b03/packages/components/toast/src/use-toast.tsx#L31)), so we can use `FormatMessage` there as well.

Any other ideas where to find those strings?

carletex commented:
Great work @dgrcode!! 

> A bit more tricky can be to find strings outside of the returned jsx. I'd look for aria-label in the whole repo. Calls to toast or useToast can also be tricky. In that case, both title and description props accept react components, so we can use FormatMessage there as well.

Toasts are important, but don't worry too much about other details like "aria-label" (at least for now)

Main components/pages are the `HomeView`, `BuilderProfileView`, `ChallengeDetailView` (even tho the README will come from Github). You can ignore `SubmissionReviewView`.

We will also need to take care of `packages/react-app/src/data/challenges.js`, which is used for display purposes (label + description. Branch could be dynamic like `base-branch-{langcode}`). What are the options? Off the top of my head:
1. Have individual `challenges_{langcode}.js` files. We'll duplicate a bunch of data.
2. Since it's a js file (and not json) we can do the i18n there, and treat it like a regular component file?

### To send to the EF translators.

- [ ] SRE front-end `en.json` file (we'll be able to generate it with this PR)
- [ ] All the READMEs from the challenges (Nothing to do in this PR)
- [ ] Autograder response. This is for another PR (even repo :)). But just noting it here: https://github.com/austintgriffith/speedrun-grader


### Next steps (future PRs?)

We also need to do all the language selection logic: the selector itself, the routing (langcode suffix on the path? /en /es). But we prolly want to merge this PR and work on the language selection logic in another PR.

Thanks!

dgrcode commented:
Regarding `packages/react-app/src/data/challenges.js`, I guess the easiest would be to have different files, one for each language (your option 1). The only problem I see with that is that translators might not be confortable creating .js files, and might prefer just touching `{lang}.json` files?

I'm going through the files you mentioned and the components imported from them. I'll be updating here as I progress:
- [x] HomeView
  - [x] ChallengeExpandedCard. I haven't touched `label` defined in line 67. Is it needed?
    - [ ] Join BG
- [ ] BuilderProfileView
- [ ] ChallengeDetailView

carletex commented:
> Regarding packages/react-app/src/data/challenges.js, I guess the easiest would be to have different files, one for each language (your option 1). The only problem I see with that is that translators might not be confortable creating .js files, and might prefer just touching {lang}.json files?

I think we should have one file and use the imperative solution there, since the object key, id, disabled, previewImage, dependencies, (even branch-name if we dynamically load the READMEs, appending the langcode). Making a change to any of these fields will imply tweaking all of them.

Not sure if `import { challengeInfo } from "../data/challenges";` will automagically pick the i18n context and give you the right translation. If that's the case, let's 100% go with a single file. If not, let's rethink the options.

> Note on this one. When the signature fails after signing with the wallet, the server might respond with an error message using error.response.data. In that case, we show that message to the user. We'd need to translate that one on the server. Is it needed?

If this is the only instance where we'd need to translate server stuff, we can skip it, to avoid setting up all the i18n stuff in the server. If I remember correctly, all the auto-grading feedback comes from https://github.com/austintgriffith/speedrun-grader. At some point, we'd need to implement server i18n there.

dgrcode commented:
Re: `packages/react-app/src/data/challenges.js`, sounds good. I'll see how it gets the context. Worst case we can initialize it somehow like adding a method like `setChallengesLang` that would set a constant in that file with the language, so we can pass it to `intl`.

As for the server i18n, so far that's the only instance I've found. I'll continue updating [this comment](https://github.com/BuidlGuidl/SpeedRunEthereum/pull/187#issuecomment-1510358708) as I progress. We can use that one as a summary, and I'll add a new comment on the PR when I update it so you know there have been progress.

dgrcode commented:
Added more texts and two notes:
> - There are some places where the logic to show some content depends on the user being an admin. I think those aren't necessary to translate, so I'll leave them untouched unless they're needed.
>
> - Similar to challenges.js, we might want to do i18n with socials.js. The only text needing i18n there are the placeholders, so we might want to skip

dgrcode commented:
Added more texts and a notes:
> - Just to double check with you that's ok to leave the role names in English (at `react-app/src/helpers/constants.js`)
>
> - Do we want to do i18n on date formats? I saw `DateWithTooltip`

carletex commented:
This is looking good @dgrcode, good job!!

> There are some places where the logic to show some content depends on the user being an admin. I think those aren't necessary to translate, so I'll leave them untouched unless they're needed.

Yep, no need.

> Similar to `challenges.js`, we might want to do i18n with `socials.js`. The only text needing i18n there are the placeholders, so we might want to skip it.

If the i18n context works correctly when importing the file, I'd use the imperative api to translate the placeholders, since it seems straight forward.

> Just to double check with you that's ok to leave the role names in English (at `react-app/src/helpers/constants.js`)

Yep, English is fine I think.

> Do we want to do i18n on date formats? I saw `DateWithTooltip`

I think **YYYY-MM-DD, HH:mm** is international enough :D Let's skip it.


Thanks Daniel!

dgrcode commented:
Awesome! Thanks! I'll update [the comment](https://github.com/BuidlGuidl/SpeedRunEthereum/pull/187#issuecomment-1510358708) with your responses to have it all in one place.

dgrcode commented:
Just realized while looking at the table, that the `DateWithTooltip` shows the relative date like "2 years ago". There's a [`FormattedRelativeTime` component](https://formatjs.io/docs/react-intl/components#formattedrelativetime), but from what I'm seeing in their examples they say at most "days". So "2 years ago" would become something like "700 days ago".

I think this is one of those things that is just not worth it translating.

dgrcode commented:
I added the last component that was remaining. I think this can be merged as a first iteration.

Next iteration would be to check how to use a specific language integrating it with the route like `speedrunethereum.com/es/...`.

Also check how to do i18n with `challenges.js` and `socials.js`.

dgrcode commented:
@carletex I've lost the button to make this a PR instead of a draft. I can only see the option to close it or comment 🤷

carletex commented:
Looking great @dgrcode 

> Also check how to do i18n with challenges.js and socials.js.

Can we test this in this PR? If it's straightforward (i18n context works correctly when importing the file) let's include it here, so we can send the `en.json` file to the translators while we work on the routing.

Thanks!

dgrcode commented:
sure! I'll look into it

dgrcode commented:
The best solution I've found is to modify `challenges.js` to return a function `getChallengeInfo` instead of the challenges object. That function receives the `intl` object, so we can do stuff like:
```js
getChallengeInfo = intl => ({
  "simple-nft-example": {
    label: intl.formatMessage({
      id: "challenges.simple-nft-example.label",
      defaultMessage: "🚩 Challenge 0: 🎟 Simple NFT Example",
    }),
    // ...
  }
})
```

I've tested it works. I'll continue adding `formatMessage` to the relevant fields. Just to double check, we only need translations for `label` and `description`, right?

carletex commented:
> I've tested it works. 

Sounds good!

>  Just to double check, we only need translations for label and description, right?

Yes!

Thanks Dani

dgrcode commented:
I've added i18n to `challenges.js`. I've removed an example button I had in a separate commit https://github.com/BuidlGuidl/SpeedRunEthereum/pull/187/commits/441875b2f938056ebc2df519b12707a278d11168 so we have a reference of how to change the locale manually.

carletex commented:
Looking great @dgrcode thanks!

I sent the `en.json` files to the translators, to see if everything looks good for them.

Could you fix the warnings, Daniel? (check the react console - `yarn start`)

![image](https://user-images.githubusercontent.com/2486142/235435873-f439b13b-7434-4b7a-b00d-535ec8bb4f84.png)

---

@technophile-04 could you help me test this? TLDR: We are preparing SRE to be translated, and this PR wraps every string with i18n functions/components. Translations are still not done, but we want to merge this PR (default to EN) so everything should be working as in `main`. While the translations are being worked, we'll be implementing the language selector / routes, so when we have all the files (es.json, hi.json, etc) we'd just add the files and enable the selector.

technophile-04 commented:
This is really cool !!! Tysm for the TLDR; also read the whole discussion, will start testing it out 🙌

dgrcode commented:
woops! I'll look into the warnings and push a commit

dgrcode commented:
done, they should be fixed now!

technophile-04 commented on packages/react-app/src/App.jsx:
```diff
@@ -42,7 +42,7 @@ const web3Modal = new Web3Modal({
 });
 
-function App() {
+function App({ setLocale }) {
```
we can remove `{ setLocale }` since its not being used

dgrcode replied:
oh yes, I forgot it there. Will remove!

technophile-04 commented:
Just tested it out nicely really nice work @dgrcode  !!!! I wasn't really able to find any big bugs/issues and functionality wise everything works 🚀. 

### Assumptions for test : 
- We are ignoring Admin side messages / pages keeping them as it is for now,
	- `SubmissionReviewView`
	- `ActivityView`
- Ignoring files / things mentioned in https://github.com/BuidlGuidl/SpeedRunEthereum/pull/187#issuecomment-1510358708 note part


### How did I test : 
For testing I generated [es.json](https://github.com/technophile-04/SpeedRunEthereum/blob/react-intl-test/packages/react-app/src/lang/es.json) and [hi.json](https://github.com/technophile-04/SpeedRunEthereum/blob/react-intl-test/packages/react-app/src/lang/hi.json) using this [cli tool](https://github.com/mololab/json-translator) (its not perfect but for testing purpose its good) and then revied both code and UI

 ### Test
- [X]  Imperative API's works nicely like in challenge.js
- [X] Functinality wise also everthing works nicely.
- [X]  Almost all the strings are converted to pass through `react-intl` leaving things mentioned in the "Assumption for test" section and I think we might have missed (not really sure on this though)
	- [ ] Navlinks 
	- [ ] Footer ("Build with" part) 


### Thought : 
This is not really needed and it just came to my mind instead passing `defaultMessage` everytime to `FormattedMessage` what if we 
if we create an custom format message component like :
```ts
import  enLang  from  "../lang/en.json";
export  default  function  CustomFormattedMessage({  stringId,  string,  values  }) {
return (
	<FormattedMessage  id={stringId}  defaultMessage={enLang[stringId]}  values={values}>
		{string}
	</FormattedMessage>
  );
}
```
Since the default is message same as `en`.  But hopefully it does not turns out bad abstraction so not too strong on this. Also `yarn translations-extract` will not work with this I think. 


Also saw GH actions is failing due to node version issue, will create a PR around it

carletex commented:
Thanks for the review @technophile-04 !!

>  Navlinks
> Footer ("Build with" part)

Yep, we definitely want to translate that too! cc @dgrcode 

> if we create an custom format message component like
>  Also yarn translations-extract will not work with this I think.

The process is creating the <FormattedMessage> (with a default message) & then run the `translations-extract`, which will take the default message and adds it to the `en.json` file. So I don't think what you are proposing would work: I think it needs "static" strings to be able to extract.

dgrcode commented:
> Yep, we definitely want to translate that too! cc @dgrcode

I'll add it to those places and push a commit!

dgrcode commented:
Something that crossed my mind as I was adding some `<FormattedMessage />`. Maybe we don't want to translate the term "builder" since it's kind of a community slang. I was thinking it might become translated as "constructor" in Spanish, which is correct by any dictionary, but it definitely has a different meaning.

What do you think @carletex?

dgrcode commented:
Something undocumented (as far as I could find): don't used dashed names for values. I updated the gotchas section.
> ## Value names with dashes
> Something like this wouldn't work:
> ```jsx
> <FormattedMessage
>   id="message.id"
>   defaultMessage="Hello {icon-heart}"
>   values={{
>     "icon-heart": <HeartFilled />,
>   }}
> ```
> But naming it `{iconHeart}` instead, does work

dgrcode commented on packages/react-app/src/components/SiteFooter.jsx:
```diff
@@ -12,17 +13,27 @@ const SiteFooter = () => {
-          </HStack>
-          <Link color={linkColor} href="https://buidlguidl.com/" isExternal>
-            BuidlGuidl
-          </Link>
+          <FormattedMessage
+            id="footer.built-with-love-at-buidlguidl"
+            defaultMessage="Built with {heartIcon} at <Link>BuidlGuidl</Link>"
+            values={{
+              heartIcon: (
+                <HStack alignItems="center" mx="5px">
+                  <HeartFilled />
+                </HStack>
+              ),
+              Link: chunks => (
+                <Link color={linkColor} href="https://buidlguidl.com/" isExternal>
+                  {chunks}
+                </Link>
+              ),
+            }}
+          />
```
Here I made a change to the html structure of the footer. I think the visual result is exactly the same, but we might want to double check it. The reason is that otherwise the translation string would have ended up being:
```
<HStack><span>Built with</span> {heartIcon} <span>at</span></Hstack><Link>BuidlGuidl</Link>
```

That html is complex enough that I don't think a typical translator could re-create it in case the order of the phrase changed in a different language.
